### PR TITLE
[12.0][IMP] base_tier_validation

### PR DESCRIPTION
**Pull request by ps-tubtim**
- Add comment on validation
- Add approve by sequence

Selection_106

1. If you check **Comment**, you can add comment on reviews table by wizard comment after 'validate' or 'reject'.
Selection_097
Selection_098

2. If you check **Approve by sequence**, approval will order by sequence.

### CODE REVIEW

**Comment by kittiu** [comment]
cc @lreficent @jbeficent 
Do you approve the proposed improvement into this module?

File: base_tier_validation/__manifest__.py
Comment (kittiu) [inline base_tier_validation/__manifest__.py]: @ps-tubtim when do the [MIG] or [IMP], we do not change the already existing code if it is not an improvement or bug fix.
(But it is our new addon, we can format whatever we want :)

File: base_tier_validation/__manifest__.py
Comment (kittiu) [inline base_tier_validation/__manifest__.py]: Here too, and all the place.

File: base_tier_validation/models/res_users.py
Comment (kittiu) [inline base_tier_validation/models/res_users.py]: No need to change other's copy right. :p

**Comment by kittiu** [comment]
@ps-tubtim I think most code is about formats, which is not necessary (it is also difficult to know what is done by you).
Please do the following first, and I will review again,
- Remove all the formatting changes, so it leave only real changes.
- Please do test script
- In the PR description, please give more detailed on what this PR try to do
- Please add the new feature in the readme.

**Comment by etobella** [comment]
Interesting approach, but just a few questions:
1. Will we be able to reference to someone like `@Administrator` on real comments?
2. It is interesting the approve level, but If you are waiting for someone, the warning on the topbar should not be shown. I don't see any js on the PR. Is it solved?

**Comment by kittiu** [comment]
> 1. Will we be able to reference to someone like `@Administrator` on real comments?

You mean @Administrator like in github that also send the email or notify? Not at the moment, as I think it is outside scope of this PR. Except it is done by some one, we can include it here.

> 2. It is interesting the approve level, but If you are waiting for someone, the warning on the topbar should not be shown. I don't see any js on the PR. Is it solved?

Agree the top bar should not show, @ps-tubtim is this possible to do?

@etobella and/but what do you mean about JS?

**Comment by etobella** [comment]
Well, probably, no JS is required, as the function that finds all the elements could be on Python, not in Javascript, sorry for the misunderstanding.

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Hi,

thanks for the proposal! here they go some issues I found testing it:

1. As already said if you have a sequence and the previous validation hasn't been done, you should not see the validation buttons (this is more in the final implementation view most likely) neither in the systray.
picture-2019-18-10_1

2. If you have sequence, you cannot validate but you can reject:
Peek 2019-10-18 16-04

3. If you open the comment wizard and you close it, the operation gets validated/rejected anyway:
Peek 2019-10-18 16-04-2

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: No need to change anything (I don't see any diff between + and - line).

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [inline base_tier_validation/wizard/comment_wizard.py]: If string is same as field name, no need to have string
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#fields

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [inline base_tier_validation/wizard/comment_wizard.py]: If string is same as field name, no need to have string
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#fields

**Comment by kittiu** [comment]
@lreficent may I ask, what tools you use to generate GIF? It is useful. :)

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [inline base_tier_validation/wizard/comment_wizard.py]: You can use self.env.ref('<actoin's external id>') to get action to save multiple lines, can you?

File: base_tier_validation/models/tier_definition.py
Comment (kittiu) [inline base_tier_validation/models/tier_definition.py]: No need for string=

File: base_tier_validation/__manifest__.py
Comment (kittiu) [inline base_tier_validation/__manifest__.py]: You don't need to bump version. As new rule, the person who merge will use ocabot to bump it.

File: base_tier_validation/__manifest__.py
Comment (kittiu) [inline base_tier_validation/__manifest__.py]: Don't need to add here. We are just doing enhancement .
But you can still add yourself as contributor in CONTRIBUTOR.rst

File: base_tier_validation/models/tier_definition.py
Comment (kittiu) [inline base_tier_validation/models/tier_definition.py]: Done need.

File: base_tier_validation/models/tier_review.py
Comment (kittiu) [inline base_tier_validation/models/tier_review.py]: Can remove.

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: I am not sure it is good idea to have logic to check and raise error in a compute field.
Is it possible to do it other way?

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: If logic is lengthy, better to use subroutine, and overall it will look nicer, i.e.,
```
rec._validate_tier()
rec._add_comment()
rec._update_counter()
```

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: Use reusable subroutine.

**Comment by LoisRForgeFlow** [comment]
> @lreficent may I ask, what tools you use to generate GIF? It is useful. :)

@kittiu I use [Peek](https://github.com/phw/peek), it is really usable and handy, you should try it :wink:

**Comment by kittiu** [comment]
@ps-tubtim
- [ ] Test case
- [ ] Readme on the new case

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: If review_ids is multi records, I am not so sure this will has singleton error.
Can you check and confirm?

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: Overall, can this logic become shorter? I looks quite complicated to me.

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py:70]: * Can len(sequence) = 0, and will sequence[0] throw error?
* Can len(sequence) > 0, and why do we use index [0]. I.e., is it sorted somehow?

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [inline base_tier_validation/wizard/comment_wizard.py]: I am a bit confused by this `view_id` variable you are using.
Are we trying to get the inheriting object? Why is it related to ir.ui.view?

File: base_tier_validation/models/tier_validation.py
Comment (kittiu) [inline base_tier_validation/models/tier_validation.py]: self.id is the view_id? I though it is tier.validation object. To me, it should be,
```suggestion
                'default_res_id': self.id,
```

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [inline base_tier_validation/wizard/comment_wizard.py]: After I try to understand it, it looks like just a res_id as integer. Nothing related to ir.ui.view, is it???
```suggestion
    res_id = fields.Integer(
```

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [reply base_tier_validation/wizard/comment_wizard.py]: If I am correct that you try to get the inheriting object, i.e., purchase.request, purchase.order, etc.
Naming variable using `action` is very misleading. You better use, i.e., object or record...

File: base_tier_validation/wizard/comment_wizard.py
Comment (kittiu) [inline base_tier_validation/wizard/comment_wizard.py]: Again, if I am not wrong (correct me if I am wrong, coz I am very confused now... O_0), how come it has external id.
```suggestion
        rec = self.env[self.res_model].browse(self.res_id)  # use rec instead of act.
```
I think we don't need to use self.env['ir.model.data'].search() at all.

File: base_tier_validation/models/tier_definition.py
Comment (kittiu) [inline base_tier_validation/models/tier_definition.py]: Does meaning of `approve_level` means to approve by `sequence` field? If so, i would change the name to,
```
approve_sequence = fields.Boolean(
    string='Approve by sequence',
    default=False,
    help="Approval order by the specified sequence number",
```
And in the view, better this field is right under field `Sequence` to imply its relationship.

File: base_tier_validation/wizard/comment_wizard.py
Comment (ps-tubtim) [reply base_tier_validation/wizard/comment_wizard.py]: I will change this code follow https://github.com/OCA/server-ux/pull/98#discussion_r341811986

File: base_tier_validation/models/tier_validation.py
Comment (ps-tubtim) [reply base_tier_validation/models/tier_validation.py]: I will change to compute field like can_review.

File: base_tier_validation/models/tier_validation.py
Comment (ps-tubtim) [reply base_tier_validation/models/tier_validation.py:70]: `sequence` has at least 1 from `if rec.can_review and rec.approve_level:`

File: base_tier_validation/models/tier_validation.py
Comment (ps-tubtim) [reply base_tier_validation/models/tier_validation.py]: Edited

File: base_tier_validation/wizard/comment_wizard.py
Comment (ps-tubtim) [reply base_tier_validation/wizard/comment_wizard.py]: Edited

File: base_tier_validation/wizard/comment_wizard.py
Comment (ps-tubtim) [reply base_tier_validation/wizard/comment_wizard.py]: Edited

File: base_tier_validation/models/tier_definition.py
Comment (ps-tubtim) [reply base_tier_validation/models/tier_definition.py]: Edited

File: base_tier_validation/models/tier_validation.py
Comment (ps-tubtim) [reply base_tier_validation/models/tier_validation.py]: `rec.reviewer_ids` and `r.reviewer_ids` are different.
- `rec.reviewer_ids`  are reviewer of document.
- `r.reviewer_ids` are reviewer of `rec.review_ids`.

**Comment by ps-tubtim** [comment]
@etobella @lreficent @kittiu Thank you for review. Please test again.

**Comment by kittiu** [comment]
@ps-tubtim please add test and more readme to greenify the codecov.

**Review by kittiu — APPROVED** [review]

**Comment by LoisRForgeFlow** [comment]
@ps-tubtim Could you update the PR to fix existing conflicts? I will review it after that! Thanks

**Comment by LoisRForgeFlow** [comment]
@ps-tubtim you should have done a rebase, you have introduced some mistakes in the code. Can you try with a rebase?

**Review by LoisRForgeFlow — APPROVED** [review]
Tested locally, looks much better! :+1:

We have been holding this for a long time, so let's move forward!

**Pull request merged**
